[V1] AsyncLLM data parallel

njhill opened the pull request:
The engine core client starts an engine core proc per dp rank and load balances requests between them. A dummy request is sent to idle ranks when the global req count goes from 0->1, and when each engine finishes all requests it will continue in an idle forward loop.

Working for single node:
```pytthon
vllm serve -dp 2 ...
```

I aimed to keep the data parallel logic isolated as much as possible (in subclasses of the core engine and client) to avoid adding complexity/overhead to the more common default dp=1 case.

Follow-on after this PR:
- Multi-node
- Balance based on waiting queue lengths rather than in-flight counts
- Make it work with API server scale-out

njhill committed:
[V1] AsyncLLM data parallel WIP

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Handle pausing loop

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
More single-node updates

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
some cleanup

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
fix up utility methods

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
revert config check

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
fixes

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
cleanup

Signed-off-by: Nick Hill <nhill@redhat.com>

youkaichao commented on vllm/v1/engine/core_client.py:
```diff
@@ -271,29 +312,23 @@ def sigusr1_handler(signum, frame):
-            })
+
+        dp_size = vllm_config.parallel_config.data_parallel_size
+        self.dp_group = None if dp_size <= 1 else (
```
do we need to check if we are doing offline inference? we will create a `dp_group` in `llm_engine` in that case. and then we should not create one here.

njhill replied:
@youkaichao actually this class currently won't be used in that case, since you enforce `VLLM_ENABLE_V1_MULTIPROCESSING` which will use `InprocClient` and bypass the zmq layer.

But I agree we should make sure that these two ways of running DP don't interfere with each other.

One question is whether offline/`LLM`-based DP will _always_ be done via your SPMD method, or if we want an alternative option to create `LLM` with `data_parallel_size=N` where it launches/manages the engines itself. I think might make less sense for multi-node though.

youkaichao replied:
in `examples/offline_inference/torchrun_example.py` , when `distributed_executor_backend="external_launcher"` , we do turn off `VLLM_ENABLE_V1_MULTIPROCESSING` , but that's not for data parallel. that's another story.

see https://github.com/vllm-project/vllm/blob/7f0be2aa24dfacf6f7aaf92411fdcd2f55ff8478/vllm/config.py#L1437

njhill replied:
Ah ok, thanks I will take a look at that. I guess `examples/offline_inference/data_parallel.py` is equivalent to the external launcher / torchrun case though right? You just happen to launch the identical procs yourself.

For the time being, I'm focused on the `AsyncMPClient` case which is still separate to the `SyncMPClient` used by LLMEngine. I'll make sure to avoid introducing any conflict in the superclass.

youkaichao replied:
> I guess examples/offline_inference/data_parallel.py is equivalent to the external launcher

no, there's a difference. 

"external_launcher" means, even if we are using one engine (one model replica), the workers are launched outside of the engine. say we use TP=2, then each engine only launches 1 workers, and work with workers launched externally.

data parallel means, we create multiple engines. inside the engine, say TP=2, the engine itself still launches 2 workers.

njhill replied:
@youkaichao fwiw I realized this particular line shouldn't be there (this is in the front-end/client proc not the engine proc), now cleaned up :)

It's already done within the engine in `core.py`.

njhill replied:
@youkaichao I have also now pushed a [change](https://github.com/vllm-project/vllm/pull/13923/commits/a4ec81b8a3b00f8eab74a1eed14f852c094b1ffe) to have LLMEngine share the new core logic in the case that it's decoupled (i.e. `VLLM_ENABLE_V1_MULTIPROCESSING=1` which is default). This should be more efficient since it doesn't wait for the client to call `step()` (the whole reason for having separate procs in the first place).

For the offline single-process case your LLMEngine logic will still be used. So the dp process group will only be initialized once in all cases. Let me know what you think!

njhill committed:
fixes

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
reconcile with LLMEngine DP in decoupled engine case

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
minor simplification

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
rework

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
class refactor

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
fix

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
adjust core engine init

Signed-off-by: Nick Hill <nhill@redhat.com>

v-lmn commented:
how to test,I mean how to run the server,I think we need two command right?
terminal 1 command 1
terminal 2 command 2
can you complete the command line

youkaichao commented on vllm/config.py:
```diff
@@ -1423,10 +1424,15 @@ def __post_init__(self) -> None:
-        self.data_parallel_master_ip = envs.VLLM_DP_MASTER_IP
-        self.data_parallel_master_port = envs.VLLM_DP_MASTER_PORT
+        if self.data_parallel_size > 1:
+            self.data_parallel_master_port = get_open_port()
```
will this function be executed again in every process, and then they have different `data_parallel_master_port` ?

njhill replied:
@youkaichao I don't think so, since this is in `__post_init__` and the config objects are passed to the subprocesses already constructed.

youkaichao reviewed (approved):
the DP-related part looks good to me.

cc @robertgshaw2-redhat I'm not familiar with the frontend processing part, maybe Robert can take a look?

njhill commented:
> how to test,I mean how to run the server,I think we need two command right?
> terminal 1 command 1
> terminal 2 command 2
> can you complete the command line

@v-lmn no for single node you can run a single command, with `--data-parallel=N`. Multi-node isn't added yet but when it is, that will require a different command to be run on the other node(s).

njhill committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into multi-engine

# Conflicts:
#	vllm/v1/core/scheduler.py
#	vllm/v1/engine/core.py
#	vllm/v1/engine/core_client.py

njhill committed:
fix new typing

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
fix :facepalm:

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
bind socket first

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
do you have to let it linger

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

# Conflicts:
#	vllm/v1/engine/core.py

njhill committed:
add comments

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
aggregate stats

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

Signed-off-by: Nick Hill <nhill@redhat.com>

# Conflicts:
#	vllm/utils.py
#	vllm/v1/engine/core_client.py

njhill committed:
Fix test

Signed-off-by: Nick Hill <nhill@redhat.com>

tlrmchlsmth commented on vllm/v1/engine/core_client.py:
```diff
@@ -504,28 +580,154 @@ async def abort_requests_async(self, request_ids: list[str]) -> None:
+                    copy=False) for engine in self.core_engines
+            ])
+
+    def get_core_engine_for_request(self) -> CoreEngine:
+        return min(self.core_engines, key=lambda e: e.num_reqs_in_flight)
```
nit: maybe factor this logic out into its own class so it's easier to swap out the scheduling logic later

njhill replied:
Thinking to leave that to a follow-on since will be changing it anyhow for the multi api-server case

tlrmchlsmth commented on vllm/v1/engine/core.py:
```diff
@@ -404,3 +426,85 @@ def process_output_socket(self, output_path: str):
+                local_unfinished_reqs)
+
+            if not self.global_unfinished_reqs:
+                # Notify client that we are pausing the loop.
+                self.output_queue.put_nowait(
+                    EngineCoreOutputs(engine_paused=True))
```
When we pause the loop, does that just mean that we're spinning in `self._process_input_queue()`? And in effect this is just telling the client to send us a `start_dp_msg` if needed?

njhill replied:
Yes, though not actually spinning because it blocks on the input socket. Really it's saying that the engine_busy_loop is going to block wait on the input queue (which is fed from the input socket by `self._process_input_queue()` once any new messages arrive.

tlrmchlsmth commented:
BTW do you know what's going on in the `v1-test`? Looks like an OOM, so I tried restarting it but it still fails. I also tried to repro it locally, but it's green for me

tlrmchlsmth reviewed (approved):
Looks good to me! Also tried it out in #13931 and it worked well for me there. 

Have you run any benchmarks? Would be good to make sure we're not causing any performance regressions in the non-DP case

njhill commented:
Thanks @tlrmchlsmth
> BTW do you know what's going on in the v1-test? Looks like an OOM, so I tried restarting it but it still fails. I also tried to repro it locally, but it's green for me

No am investigating, I expect it is somehow caused by this though :( usually those kind of OOMs are where proc from previous test hasn't shut down properly.

> Have you run any benchmarks? Would be good to make sure we're not causing any performance regressions in the non-DP case

No but I will do that too, also need some CI test coverage...

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill committed:
fix and minor cleanup

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Add CI test

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill commented:
Performance tests show a possible slight impact. It looks negligible but the difference is pretty consistent. I'm investigating.

Before
```
============ Serving Benchmark Result ============
Successful requests:                     6000      
Benchmark duration (s):                  88.80     
Total input tokens:                      1322102   
Total generated tokens:                  1205048   
Request throughput (req/s):              67.57     
Output token throughput (tok/s):         13569.85  
Total Token throughput (tok/s):          28457.83  
---------------Time to First Token----------------
Mean TTFT (ms):                          236.66    
Median TTFT (ms):                        195.71    
P99 TTFT (ms):                           1273.58   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          28.41     
Median TPOT (ms):                        28.19     
P99 TPOT (ms):                           41.10     
---------------Inter-token Latency----------------
Mean ITL (ms):                           29.43     
Median ITL (ms):                         27.59     
P99 ITL (ms):                            53.60     
==================================================
```
After
```
============ Serving Benchmark Result ============
Successful requests:                     6000      
Benchmark duration (s):                  89.40     
Total input tokens:                      1322102   
Total generated tokens:                  1205048   
Request throughput (req/s):              67.12     
Output token throughput (tok/s):         13479.51  
Total Token throughput (tok/s):          28268.36  
---------------Time to First Token----------------
Mean TTFT (ms):                          238.52    
Median TTFT (ms):                        202.93    
P99 TTFT (ms):                           1261.84   
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          28.52     
Median TPOT (ms):                        28.25     
P99 TPOT (ms):                           41.31     
---------------Inter-token Latency----------------
Mean ITL (ms):                           29.53     
Median ITL (ms):                         27.58     
P99 ITL (ms):                            53.41     
==================================================
```

njhill committed:
Some simplification and fixes

Signed-off-by: Nick Hill <nhill@redhat.com>

markmc reviewed (commented):
Note sure how useful these comments are ...

And maybe I'm blinded by the recent parallel sampling refactor, but my hunch is that the multi-core-engine handling might be cleaner at the `AsyncLLM` level - have a list of `EngineCoreClient`s, do the output processor stuff (like tracking in-flight requests) in `AsyncLLM`'s output processor, have all the stats aggregation logic at that level ... I'd need to spend more time on it to feel strongly about it though :+1:

tlrmchlsmth commented on vllm/engine/arg_utils.py:
```diff
@@ -439,6 +440,11 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
                             type=int,
                             default=EngineArgs.tensor_parallel_size,
                             help='Number of tensor parallel replicas.')
+        parser.add_argument('--data-parallel-size',
+                            '-dp',
+                            type=int,
+                            default=EngineArgs.data_parallel_size,
+                            help='Number of data parallel replicas.')
```
Could you add to the help string that MoE layers will be sharded according to the product of DP and TP?

Reference discussion here https://github.com/vllm-project/vllm/pull/14305#discussion_r1983588471

njhill commented on vllm/v1/engine/__init__.py:
```diff
@@ -134,6 +134,15 @@ class EngineCoreOutputs(
+    # Set to False to indicate stats should be accumulated rather than
+    # recorded, when there are remaining outputs from other engines
+    # still to come for this iteration.
+    final_outputs_for_step: bool = True
```
Yes you're right and I also don't like it and had been thinking about alternatives. For this first PR I had been trying to avoid changing too much.

It's a struct from core->client comm but also what's returned from the client. Agree that we could / probably should separate these things. But TBH I'm not set on this mechanism for aggregating the metrics so was expecting it may be reworked anyhow.

njhill commented on vllm/v1/core/scheduler.py:
```diff
@@ -583,10 +585,14 @@ def update_from_output(
         )
+        if self.include_finished_set:
+            engine_core_outputs.finished_requests = (
+                scheduler_output.finished_req_ids)
```
> Are these different from the outputs with a non-None `finish_reason` ? If so, we're duplicating this information because we don't want the core client to loop over all outputs (we do that at the AsyncLLM level)?

That's correct, and these finished req ids will be "sparse" i.e. only one per request, so the overhead should be minimal. But the alternative would mean looping through all requests in the batch on every step (in the client).

njhill commented on vllm/v1/engine/core.py:
```diff
@@ -402,4 +427,94 @@ def process_output_socket(self, output_path: str):
+
+
+ENGINE_PAUSED_OUTPUTS = EngineCoreOutputs(engine_paused=True)
+EMPTY_STATS_OUTPUTS = EngineCoreOutputs(scheduler_stats=SchedulerStats())
```
Agree, this is also something I had been thinking about! The main reason I didn't (yet) is that I was trying to limit the scope of this PR. If it's not added in this one we should do as a follow-on.

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -252,45 +254,49 @@ async def _run_output_handler(self):
         """Background loop: pulls from EngineCore and pushes to AsyncStreams."""
 
         try:
+            iteration_stats = IterationStats() if self.log_stats else None
```
I'm not sure :) I was thinking that they are, because the scope is still a single iteration across all of the engines. In terms of how the stats are aggregated/reported this seemed like the right thing to do. Otherwise we'd need to need to generalize the metrics reporting to include an engine label (though maybe this wouldn't be a bad idea and would potentially be simpler in that we wouldn't need to aggregate here).

For the logged stats though, I was thinking that it's still most useful to show the aggregate numbers. That's probably what you want if you are just launching an AsyncLLM .. it's a implementation detail to some extent that there's one or multiple engines behind it.

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -306,19 +312,22 @@ async def abort(self, request_id: str) -> None:
+    def _add_stats(self, scheduler_stats: Optional[SchedulerStats]):
+        if scheduler_stats:
+            if self.scheduler_stats is None:
+                self.scheduler_stats = scheduler_stats
```
I was mostly conforming to what was already there ... the local iteration_stats var is already get "reused" over multiple passes of `process_outputs` and this is just extending the scope of that. Accumulating the scheduler_stats in a member kept that logic out of the main loop.

> Maybe just include `scheduler_stats` in `iteration_stats`?

Good idea, let me try that!

njhill commented on vllm/v1/engine/core_client.py:
```diff
@@ -444,13 +496,18 @@ def __init__(self, vllm_config: VllmConfig, executor_class: type[Executor],
         self.queue_task: Optional[asyncio.Task] = None
 
+        self.output_processor: Optional[Callable[
+            [AsyncMPClient, EngineCoreOutputs], Awaitable[None]]] = None
```
lol yeah I actually hadn't noticed that, wasn't thinking about the "other" output_processor when adding this. I'll rename it.

njhill commented on vllm/v1/engine/core_client.py:
```diff
@@ -463,7 +520,16 @@ async def process_outputs_socket():
+                    assert _self_ref is not None
+                    _self = _self_ref()
+                    if not _self:
+                        return
```
Yeah it's kind of nasty but needed to avoid the circular ref so that garbage collection isn't inhibited. Once that returns None then the whole client is gone and the task should no longer be running. Will add a comment :+1:

njhill commented on vllm/v1/engine/core_client.py:
```diff
@@ -504,28 +578,153 @@ async def abort_requests_async(self, request_ids: list[str]) -> None:
+                if coros:
+                    await asyncio.gather(*coros)
+
+        if outputs.scheduler_stats:
```
Yes that's right. When there are no stats being logged/propagated then this doesn't apply and so that logic doesn't need to run.

njhill commented on vllm/v1/engine/async_llm.py:
```diff
@@ -252,45 +254,49 @@ async def _run_output_handler(self):
-                )
+                if self.log_stats:
+                    self._add_stats(outputs.scheduler_stats)
+                    if outputs.final_outputs_for_step:
```
When we run the model forward pass for DP, all engines have to participate because the experts are shared across them in the MoE layers. So even when only one engine is processing requests, the others will still be looping.

Right now, to make this counting logic work, I have these empty loops returning EMPTY_SCHED_STATS each time, to ensure there are always N sets of stats returned per global step (currently, if that wasn't done, this flag would still get set every N stats received but the grouping would be wrong i.e. over multiple steps for the same engine).

However, as alluded to in my [other](https://github.com/vllm-project/vllm/pull/13923#discussion_r1983715404) response I really don't like relying on this counting approach and would like to replace it with something where we don't need to send back empty stats from the "idle" looping engines.

It's non-trivial though since I've tried to minimize the amount of synchronization happening and it's not easy to know when a response comes back from a particular engine whether there are other engines that participated in the same step (i.e. whether to wait for those responses before aggregating). Partly because there are no guarantees of the order of responses from a particular step, and each engine doesn't know about others that may or may not have processed requests in that step.

njhill commented:
Thanks @markmc for the great comments!

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

# Conflicts:
#	vllm/engine/arg_utils.py

njhill commented:
> And maybe I'm blinded by the recent parallel sampling refactor, but my hunch is that the multi-core-engine handling might be cleaner at the `AsyncLLM` level - have a list of `EngineCoreClient`s, do the output processor stuff (like tracking in-flight requests) in `AsyncLLM`'s output processor, have all the stats aggregation logic at that level ... I'd need to spend more time on it to feel strongly about it though 👍

@markmc I forgot to respond to this. I will also think about it some more but my initial feeling is that it's cleaner to keep this encapsulated in the engine client. There's no reason that that AsyncLLM layer "needs" to know about the multi-engine case so this keeps more of the code simpler for the vastly more common non-DP case. That's also why I tried to keep most of the DP-specific logic confined to subclasses. This might change though when adding support for api-server scale-out and in particular supporting the many-many case.

njhill committed:
address @markmc's stats suggestion

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
address @tms's arg comment

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
fix utility method breakage

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
rename AsyncMPClient output_processor to output_handler

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

aarnphm commented on vllm/v1/engine/core_client.py:
```diff
@@ -197,28 +201,72 @@ def pin_lora(self, lora_id: int) -> bool:
         return self.engine_core.pin_lora(lora_id)
 
 
+class CoreEngine:
```
qq:

I felt like this is a bit confusing given that we have `EngineCore` from the "core" engine, and this `CoreEngine` acts like DPEngineCore iiuc?

njhill replied:
@aarnphm yes I don't disagree .. I'll think about how to rename/simplify. This class represents one of the engines from the client side so it's not the engine itself. It wraps the process management for that engine for the local engines (which is the case for all of them currently since only single-node is done) but am planning to generalize such that there will be subclasses for local and remote.

aarnphm replied:
makes sense. from the looks of it, it acts as a EngineCoreProxy?

njhill replied:
It's a proxy in some sense but not in the sense that you would think of when you see "proxy" :)

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

Signed-off-by: Nick Hill <nhill@redhat.com>

# Conflicts:
#	vllm/v1/core/scheduler.py
#	vllm/v1/engine/async_llm.py
#	vllm/v1/engine/core.py
#	vllm/v1/metrics/loggers.py

njhill committed:
Fix

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented on vllm/v1/engine/core_client.py:
```diff
@@ -197,28 +201,72 @@ def pin_lora(self, lora_id: int) -> bool:
 
+    core_engines: list[CoreEngine] = field(default_factory=list)
     ctx: Union[zmq.Context] = None
     output_socket: Union[zmq.Socket, zmq.asyncio.Socket] = None
```
@ZhongYingMatrix could you elaborate? I'm not sure what you mean, there's a single output_socket shared by all of the engines (but they each have their own input_socket).

ZhongYingMatrix replied:
I see..I misunderstood it. Thx!

DarkLight1337 commented:
LoRA test keeps hanging, PTAL.

njhill commented:
FYI @markmc I changed the metrics logic now in this PR to not attempt to aggregate from multiple engines and just log them separately. For prometheus I added a label so they can be aggregated externally. This avoids a lot of complication and is actually probably more useful to see the split between engines.

markmc commented:
> FYI @markmc I changed the metrics logic now in this PR to not attempt to aggregate from multiple engines and just log them separately. For prometheus I added a label so they can be aggregated externally. This avoids a lot of complication and is actually probably more useful to see the split between engines.

That actually makes a ton of sense, I think. Nice insight! e.g. the Grafana dashboard already aggregates across all instances with the same `model_name`, so it will aggregate across engines naturally

Doesn't that mean these `stats.add()` methods are unused now and can be removed?

njhill commented:
> 
> Doesn't that mean these `stats.add()` methods are unused now and can be removed?

Probably yes, thanks!

njhill committed:
Remove redundant logic related to removed stats aggregation

Signed-off-by: Nick Hill <nhill@redhat.com>

DarkLight1337 commented:
Can you merge from latest main to pass the CI?

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill committed:
Fixes

Signed-off-by: Nick Hill <nhill@redhat.com>

aarnphm commented:
qq: Do you think we can merge this for 0.8.0?

DarkLight1337 commented:
The distributed tests are failing consistently.

njhill commented:
> The distributed tests are failing consistently. Possibly due to upgrading to torch 2.6.0?

@DarkLight1337 That's right, hopefully fixed in https://github.com/vllm-project/vllm/pull/14870. Will resolve the conflicts once that's merged.

> qq: Do you think we can merge this for 0.8.0?

@aarnphm no, we aren't including this in 0.8.0.

njhill committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into multi-engine

# Conflicts:
#	.buildkite/test-pipeline.yaml

njhill committed:
fix issue from main merge

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
remove leftover unused field

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Fix offline DP compatibility

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill commented:
@hasB4K yes multi-node isn't in this PR, I am working on that as a follow-on but would like to get this merged first to avoid it becoming much bigger.

The other main thing I am digging into here is performance related. When there are multiple engines behind a single API server, the front-end process becomes a bottleneck, which negates the potential benefits of doing DP in the first place. So I am trying to address that - it may require the API-server process scale-out capability which we're planning to do regardless.

njhill committed:
Add timeout to data_parallel.py

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into multi-engine

njhill commented:
Performance now looks good with this PR for dp=1 (no regression, possibly even slightly improved):

300 concurrent requests, llama-1b on A100

main
```
============ Serving Benchmark Result ============
Successful requests:                     6000      
Benchmark duration (s):                  80.55     
Total input tokens:                      1322102   
Total generated tokens:                  1205048   
Request throughput (req/s):              74.49     
Output token throughput (tok/s):         14960.69  
Total Token throughput (tok/s):          31374.60  
---------------Time to First Token----------------
Mean TTFT (ms):                          160.41    
Median TTFT (ms):                        141.51    
P99 TTFT (ms):                           881.77    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          19.47     
Median TPOT (ms):                        19.02     
P99 TPOT (ms):                           30.88     
---------------Inter-token Latency----------------
Mean ITL (ms):                           21.76     
Median ITL (ms):                         20.76     
P99 ITL (ms):                            40.48     
==================================================
```

This PR:
```
============ Serving Benchmark Result ============
Successful requests:                     6000      
Benchmark duration (s):                  79.48     
Total input tokens:                      1322102   
Total generated tokens:                  1205048   
Request throughput (req/s):              75.49     
Output token throughput (tok/s):         15160.71  
Total Token throughput (tok/s):          31794.07  
---------------Time to First Token----------------
Mean TTFT (ms):                          158.76    
Median TTFT (ms):                        138.96    
P99 TTFT (ms):                           894.08    
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms):                          19.23     
Median TPOT (ms):                        18.77     
P99 TPOT (ms):                           31.68     
---------------Inter-token Latency----------------
Mean ITL (ms):                           21.21     
Median ITL (ms):                         20.32     
P99 ITL (ms):                            41.11     
==================================================
```

Still looking into performance with dp > 1 but that doesn't necessarily need to block this PR from being merged.

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill committed:
Enable less-frequent all-reduce optimization

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill committed:
clean distributed shutdown

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
address misc loose-ends

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

# Conflicts:
#	vllm/v1/core/sched/scheduler.py

njhill committed:
further tweaks

Signed-off-by: Nick Hill <nhill@redhat.com>

robertgshaw2-redhat commented on vllm/v1/core/sched/scheduler.py:
```diff
@@ -645,10 +647,16 @@ def update_from_output(
             outputs=outputs,
             scheduler_stats=self.make_stats(),
         )
+        if self.include_finished_set:
```
I think that `include_finished_set` warrants a more detailed comment about why / what this is doing

gglin001 commented:
hi, this pr seems have conflict with

- https://github.com/vllm-project/vllm/pull/14899

https://github.com/vllm-project/vllm/pull/14899/files#diff-cf486d32b11454e89103000ac0f9a93ee1ccbb8983c02e794b734a65301cd98b

youkaichao commented:
> hi, this pr seems have conflict with
> 
> * [[Bugfix] torchrun compatibility #14899](https://github.com/vllm-project/vllm/pull/14899)
> 
> [#14899 (files)](https://github.com/vllm-project/vllm/pull/14899/files#diff-cf486d32b11454e89103000ac0f9a93ee1ccbb8983c02e794b734a65301cd98b)

I'm fixing it in https://github.com/vllm-project/vllm/pull/15355

njhill committed:
Merge remote-tracking branch 'refs/remotes/origin/main' into multi-engine

# Conflicts:
#	.buildkite/test-pipeline.yaml
#	vllm/v1/engine/core_client.py

njhill committed:
Additional debug

Signed-off-by: Nick Hill <nhill@redhat.com>

tlrmchlsmth commented on tests/v1/test_async_llm_dp.py:
```diff
@@ -0,0 +1,108 @@
+from vllm.v1.engine.async_llm import AsyncLLM
+from vllm.v1.engine.core_client import DPAsyncMPClient
+
+if not current_platform.is_cuda():
+    pytest.skip(reason="V1 currently only supported on CUDA.",
+                allow_module_level=True)
```
We could actually use `supports_v1` now that this PR has landed (probably only want to turn tests on for CUDA and RoCM though)

https://github.com/vllm-project/vllm/pull/15417

tlrmchlsmth reviewed (approved):
Left a few minor comments. The PR still looks very good and would be great to get it landed soon.

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

Signed-off-by: Nick Hill <nhill@redhat.com>

# Conflicts:
#	examples/offline_inference/data_parallel.py

njhill committed:
Address review comments on tests

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

njhill commented:
Thanks @tlrmchlsmth! Have addressed those comments. Also had to make some additional adjustments to ensure compatibility with @youkaichao's offline multi-node scenario added in https://github.com/vllm-project/vllm/pull/15484.

njhill committed:
Fix env var fallback

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Fix test supports_v1 check

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Fix yapf :facepalm:

Signed-off-by: Nick Hill <nhill@redhat.com>

njhill committed:
Merge remote-tracking branch 'origin/main' into multi-engine

# Conflicts:
#	vllm/v1/core/sched/scheduler.py

njhill commented on vllm/v1/engine/core.py:
```diff
@@ -443,5 +486,114 @@ def process_output_socket(self, output_path: str):
+        local_dp_rank = vllm_config.parallel_config.data_parallel_rank_local
+
+        assert dp_size > 1
+        assert 0 <= local_dp_rank <= dp_rank < dp_size
```
It's not strictly needed, I just thought it might be good here to verify that the config is in a coherent state.

youkaichao commented:
@Co1lin I tested `vllm serve meta-llama/Llama-3.2-1B-Instruct -dp 2 --port 8989` on the main branch, it works well.

can you create a separate issue with detailed environment?

liunn01 commented:
vllm serve /local/models/Qwen2___5-32B -dp 8 --dtype half ，卡在图片所示不动，该如何正确执行呢
![1111](https://github.com/user-attachments/assets/df548295-5e63-45ff-9ced-4d617e1d1e70)

hidva commented on vllm/v1/engine/core.py:
```diff
@@ -443,5 +486,114 @@ def process_output_socket(self, output_path: str):
+            local_unfinished_reqs = self.scheduler.has_unfinished_requests()
+
+            if local_unfinished_reqs:
+                # 2) Step the engine core.
```
Fixed in https://github.com/vllm-project/vllm/pull/18559
